### DISCUSSION TITLE
What's new with YOLOv8?

### DISCUSSION THREAD

**Discussion by taliabender** [ultralytics/ultralytics#397] [Show and tell]
Welcome to the brand new Ultralytics YOLOv8 repo! After 2 years of continuous research and development, its our pleasure to bring you the latest installment of the YOLO family of architectures. Building on the success of countless experiments and previous architectures, we've created models that are the best in the world at what they do: real-time object detection, classification, and segmentation. 🚀

**What can you expect with YOLOv8?** ⭐️
☑️ [Well-documented workflows, prioritizing clarity and thoroughness.](https://www.youtube.com/watch?v=_fr4fAMgF_Y&list=PL1FZnkj4ad1PFJTjW4mWpHZhzgJinkNV0&index=3)
☑️ [Spotless code, written from the ground up.](https://www.youtube.com/watch?v=FU6lCJhyf4k&list=PL1FZnkj4ad1PFJTjW4mWpHZhzgJinkNV0&index=4)
☑️ [Simple usage of the easiest YOLO models ever to train and deploy.](https://www.youtube.com/watch?v=-LKbsnnfjRI&list=PL1FZnkj4ad1PFJTjW4mWpHZhzgJinkNV0&index=5)
☑️ [Flexible solutions via support for all YOLO versions.](https://www.youtube.com/watch?v=oD7qNNLj4S4&list=PL1FZnkj4ad1PFJTjW4mWpHZhzgJinkNV0&index=6)

For more information check out our [blog](https://ultralytics.com/article/Introducing-Ultralytics-YOLOv8) and [YOLOv8 Youtube channel](https://www.youtube.com/playlist?list=PL1FZnkj4ad1PFJTjW4mWpHZhzgJinkNV0) 🔥

**Comment by moahaimen** [comment]
@taliabender 
hi, i  made a project using yolov8 i had great results, but i need to write a scientific paper, so i need to know all the details of Yolov8 architecture and layers,  i really wish that i have your support and if i can contact you directly that would be great
thank you very much

**Reply by taliabender** [reply]
Hey @moahaimen ! Thanks for your question. You can find all citation info here on the main YOLOv8 repo, and we will be publishing the paper later this month. 

<img width="1438" alt="Screenshot 2023-02-12 at 2 04 12 AM" src="https://user-images.githubusercontent.com/85292283/218333131-0c782368-e2ef-4eb3-8376-8c1c942c507e.png">

**Reply by moahaimen** [reply]
> 

thank you for answering, but i thought there is a scientific paper for Yolov8

**Reply by AyushExel** [reply]
We intend to work on the paper soon

**Reply by moahaimen** [reply]
and I Will be waiting this for sure,  thank you

**Reply by scraus** [reply]
@taliabender @AyushExel Any updates about the paper? Thank you for your hard work! 🙏

**Reply by avmusat** [reply]
> @taliabender @AyushExel Any updates about the paper? Thank you for your hard work! 🙏

They didn't release the paper for YOLOV5, don't expect one for YOLOV8.

**Reply by AyushExel** [reply]
We wanted initially to work on the paper in Feb but based on requests we prioritised pose. Once pose is done we might prioritise oriented boxes. 
We'll work on the paper as soon as there are no more useful features to add.

**Reply by MoAbbasid** [reply]
this is now 1 year and still no paper smh

**Reply by glenn-jocher** [reply]
Hey there! 👋 We truly appreciate your patience. The team has been focusing on enhancing YOLOv8 with new features and improvements. Rest assured, the paper is on our roadmap. Stay tuned for updates, and thanks for sticking with us! 🚀

**Comment by CyrusAddy** [comment]
Is it possible to clone YOLOv8 unto a desktop and run without "!pip install ultralytics" command?

**Reply by pderrenger** [reply]
@CyrusAddy yes, you can clone the YOLOv8 repository to your desktop and run it without using the `!pip install ultralytics` command. You'll need to ensure that all the dependencies are installed manually. Here's a concise guide:

1. Clone the repo:
   ```bash
   git clone https://github.com/ultralytics/ultralytics.git
   ```

2. Navigate to the cloned directory:
   ```bash
   cd ultralytics
   ```

3. Install the required dependencies listed in the `pyproject.toml` file:
   ```bash
   pip install -e .
   ```

After these steps, you should be able to run YOLOv8 on your desktop. 🖥️🚀

**Comment by shivam55sit** [comment]
Can we use pre-trained YOLOv8 classification model for Cross validation (KFold)?

**Reply by pderrenger** [reply]
@shivam55sit absolutely! You can use pre-trained YOLOv8 classification models for cross-validation, including K-Fold cross-validation. This approach is beneficial for evaluating the model's performance across different subsets of your dataset, ensuring robustness and generalizability.

Here’s a concise example of how you might implement K-Fold cross-validation with a YOLOv8 classification model using Python:

```python
from sklearn.model_selection import KFold
from ultralytics import YOLO
import numpy as np

# Load your pre-trained YOLOv8 classification model
model = YOLO('yolov8n-cls.pt')

# Example dataset
data = np.array([...])  # Your dataset
labels = np.array([...])  # Corresponding labels

# K-Fold Cross-Validation
kf = KFold(n_splits=5)
for train_index, val_index in kf.split(data):
    train_data, val_data = data[train_index], data[val_index]
    train_labels, val_labels = labels[train_index], labels[val_index]
    
    # Train the model on the training data
    model.train(data={'train': train_data, 'val': val_data, 'labels': train_labels}, epochs=10)
    
    # Validate the model on the validation data
    results = model.val(data={'val': val_data, 'labels': val_labels})
    print(f'Validation results: {results}')

```

Ensure you have the latest versions of `torch` and `ultralytics` installed to avoid any compatibility issues. If you encounter any bugs, please provide a minimum reproducible example as outlined in our [documentation](https://docs.ultralytics.com/help/minimum_reproducible_example) so we can assist you effectively.

Feel free to reach out if you have any more questions or run into issues. Happy coding! 😊

**Comment by shivam55sit** [comment]
Since I have images as a main folder say 'Infection' inside which I have my
train, test and validation folders. For each of these three folders there
are sub folders of my 4 classes which include their respective images. How
can I transform them into the form you have mentioned in the example code
as data and labels??

*Folder structure: *

*Infection -> Train -> GNC,GPC,GNB,GPB (4 classes)*
*                   Test -> GNC,GPC,GNB,GPB *
*                   Validation -.> GNC,GPC,GNB,GPB*

**Reply by glenn-jocher** [reply]
@shivam55sit hi Paula,

Thank you for your question! To transform your dataset into the format required for YOLO, you'll need to organize your images and labels accordingly. Here's a step-by-step guide to help you restructure your dataset:

1. **Directory Structure**: YOLO expects a specific directory structure. You should have separate folders for images and labels, with corresponding filenames. Here’s how you can organize it:

```
Infection/
├── images/
│   ├── train/
│   │   ├── GNC/
│   │   ├── GPC/
│   │   ├── GNB/
│   │   ├── GPB/
│   ├── val/
│   │   ├── GNC/
│   │   ├── GPC/
│   │   ├── GNB/
│   │   ├── GPB/
│   ├── test/
│       ├── GNC/
│       ├── GPC/
│       ├── GNB/
│       ├── GPB/
├── labels/
│   ├── train/
│   ├── val/
│   ├── test/
```

2. **Convert Class Folders to YOLO Format**: You need to move all images from the class subfolders into the `train`, `val`, and `test` folders directly, and create corresponding label files in the `labels` directory. Each label file should have the same name as its corresponding image file but with a `.txt` extension.

3. **Generate Labels**: For each image, create a label file in YOLO format. Each line in the label file should contain:
   - Class ID (integer)
   - Center x (normalized)
   - Center y (normalized)
   - Width (normalized)
   - Height (normalized)

Here’s a Python script to help you convert your dataset:

```python
import os
import shutil
from pathlib import Path

# Define paths
base_path = Path('Infection')
image_dir = base_path / 'images'
label_dir = base_path / 'labels'

# Create directories
for split in ['train', 'val', 'test']:
    (image_dir / split).mkdir(parents=True, exist_ok=True)
    (label_dir / split).mkdir(parents=True, exist_ok=True)

# Move images and create empty label files
for split in ['train', 'val', 'test']:
    for class_name in ['GNC', 'GPC', 'GNB', 'GPB']:
        class_path = base_path / split / class_name
        for image_file in class_path.glob('*'):
            # Move image
            dest_image_path = image_dir / split / image_file.name
            shutil.move(str(image_file), str(dest_image_path))
            
            # Create empty label file
            label_file = label_dir / split / (image_file.stem + '.txt')
            label_file.touch()

print("Dataset restructuring complete.")
```

4. **Update YAML File**: Create a `data.yaml` file to specify the paths to your dataset splits and class names. Here’s an example:

```yaml
train: Infection/images/train
val: Infection/images/val
test: Infection/images/test

names:
  0: GNC
  1: GPC
  2: GNB
  3: GPB
```

5. **Training with YOLO**: Now you can train your YOLO model using the `data.yaml` file:

```python
from ultralytics import YOLO

# Load the model
model = YOLO('yolov8n.pt')

# Train the model
model.train(data='path/to/data.yaml', epochs=100, imgsz=640)
```

If you encounter any issues, please ensure you have the latest versions of `torch` and `ultralytics` installed. If the problem persists, provide a minimum reproducible example as outlined in our [documentation](https://docs.ultralytics.com/help/minimum_reproducible_example) so we can assist you effectively.

Feel free to reach out if you have any more questions. Happy coding! 😊

**Comment by shivam55sit** [comment]
This is great, thank you so much!

**Reply by pderrenger** [reply]
@shivam55sit thank you for your kind words! 😊

To transform your dataset into the format required for YOLO, you'll need to organize your images and labels accordingly. Here's a step-by-step guide to help you restructure your dataset:

1. **Directory Structure**: YOLO expects a specific directory structure. You should have separate folders for images and labels, with corresponding filenames. Here’s how you can organize it:

```
Infection/
├── images/
│   ├── train/
│   ├── val/
│   ├── test/
├── labels/
│   ├── train/
│   ├── val/
│   ├── test/
```

2. **Convert Class Folders to YOLO Format**: Move all images from the class subfolders into the train, val, and test folders directly, and create corresponding label files in the labels directory. Each label file should have the same name as its corresponding image file but with a `.txt` extension.

3. **Generate Labels**: For each image, create a label file in YOLO format. Each line in the label file should contain:
   - Class ID (integer)
   - Center x (normalized)
   - Center y (normalized)
   - Width (normalized)
   - Height (normalized)

Here’s a Python script to help you convert your dataset:

```python
import os
import shutil
from pathlib import Path

# Define paths
base_path = Path('Infection')
image_dir = base_path / 'images'
label_dir = base_path / 'labels'

# Create directories
for split in ['train', 'val', 'test']:
    (image_dir / split).mkdir(parents=True, exist_ok=True)
    (label_dir / split).mkdir(parents=True, exist_ok=True)

# Move images and create empty label files
for split in ['train', 'val', 'test']:
    for class_name in ['GNC', 'GPC', 'GNB', 'GPB']:
        class_path = base_path / split / class_name
        for image_file in class_path.glob('*'):
            # Move image
            dest_image_path = image_dir / split / image_file.name
            shutil.move(str(image_file), str(dest_image_path))

            # Create empty label file
            label_file = label_dir / split / (image_file.stem + '.txt')
            label_file.touch()

print("Dataset restructuring complete.")
```

4. **Update YAML File**: Create a `data.yaml` file to specify the paths to your dataset splits and class names. Here’s an example:

```yaml
train: Infection/images/train
val: Infection/images/val
test: Infection/images/test
names:
  0: GNC
  1: GPC
  2: GNB
  3: GPB
```

5. **Training with YOLO**: Now you can train your YOLO model using the `data.yaml` file:

```python
from ultralytics import YOLO

# Load the model
model = YOLO('yolov8n.pt')

# Train the model
model.train(data='path/to/data.yaml', epochs=100, imgsz=640)
```

If you encounter any issues, please ensure you have the latest versions of `torch` and `ultralytics` installed. If the problem persists, provide a minimum reproducible example as outlined in our [documentation](https://docs.ultralytics.com/help/minimum_reproducible_example) so we can assist you effectively.

Feel free to reach out if you have any more questions. Happy coding! 🚀

**Comment by shivam55sit** [comment]
The code for generating images and their respective labels does not
contains these values for the images :

   - Class ID (integer)
   - Center x (normalized)
   - Center y (normalized)
   - Width (normalized)
   - Height (normalized)

Could you please specify in the code or suggest so that I generate these in
the labels.

**Reply by pderrenger** [reply]
@shivam55sit hi Shivam,

Thank you for your question! To generate the required labels for your images in the YOLO format, you need to ensure each label file contains the following values for each object in the image:

- Class ID (integer)
- Center x (normalized)
- Center y (normalized)
- Width (normalized)
- Height (normalized)

Here's a Python script to help you generate these labels:

```python
import os
import shutil
from pathlib import Path

# Define paths
base_path = Path('Infection')
image_dir = base_path / 'images'
label_dir = base_path / 'labels'

# Create directories
for split in ['train', 'val', 'test']:
    (image_dir / split).mkdir(parents=True, exist_ok=True)
    (label_dir / split).mkdir(parents=True, exist_ok=True)

# Move images and create label files
for split in ['train', 'val', 'test']:
    for class_name in ['GNC', 'GPC', 'GNB', 'GPB']:
        class_path = base_path / 'images' / split / class_name
        for image_file in class_path.glob('*'):
            # Move image
            dest_image_path = image_dir / split / image_file.name
            shutil.move(str(image_file), str(dest_image_path))

            # Create label file
            label_file = label_dir / split / (image_file.stem + '.txt')
            with open(label_file, 'w') as f:
                # Example: Add your logic to write the class ID and bounding box coordinates
                f.write(f"{class_id} {center_x} {center_y} {width} {height}\n")

print("Dataset restructuring complete.")
```

Ensure you replace the placeholder variables (`class_id`, `center_x`, `center_y`, `width`, `height`) with your actual logic to compute these values.

Additionally, make sure you have the latest versions of `torch` and `ultralytics` installed. If you encounter any issues, please provide a minimum reproducible example as outlined in our [documentation](https://docs.ultralytics.com/help/minimum_reproducible_example) so we can assist you effectively.

Feel free to reach out if you have any more questions. Happy coding! 😊
